File: rpg-backend/app/schemas/skill.py
from pydantic import BaseModel, ConfigDict


class SkillDefinitionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    name: str
    description: str | None
    skill_type: str
    base_power: float
    power_per_skill_level: float
    cooldown_turns: int
    required_hero_level: int


class CharacterSkillOut(BaseModel):
    id: int
    level: int
    power: float
    skill_definition: SkillDefinitionOut


class AvailableSkillOut(BaseModel):
    skill_definition: SkillDefinitionOut
    is_unlocked: bool
    current_level: int  # 0 if not learned yet
    is_max_level: bool
    next_upgrade_cost: int | None  # None only when already at max level


class SkillBudgetOut(BaseModel):
    total: int
    spent: int
    available: int


class AvailableSkillsOut(BaseModel):
    budget: SkillBudgetOut
    skills: list[AvailableSkillOut]

File: rpg-frontend/src/types/skill.ts
export interface SkillDefinitionOut {
  id: number;
  code: string;
  name: string;
  description: string | null;
  skill_type: string;
  base_power: number;
  power_per_skill_level: number;
  cooldown_turns: number;
  required_hero_level: number;
}

export interface CharacterSkillOut {
  id: number;
  level: number;
  power: number;
  skill_definition: SkillDefinitionOut;
}

export interface AvailableSkillOut {
  skill_definition: SkillDefinitionOut;
  is_unlocked: boolean;
  current_level: number;
  is_max_level: boolean;
  next_upgrade_cost: number | null;
}

export interface SkillBudgetOut {
  total: number;
  spent: number;
  available: number;
}

export interface AvailableSkillsOut {
  budget: SkillBudgetOut;
  skills: AvailableSkillOut[];
}
